File: backend/src/main/java/com/erp/modules/purchases/service/PurchaseRequisitionService.java
package com.erp.modules.purchases.service;

import com.erp.modules.purchases.domain.dto.ConvertRequisitionRequest;
import com.erp.modules.purchases.domain.dto.CreatePurchaseRequisitionRequest;
import com.erp.modules.purchases.domain.dto.PurchaseRequisitionDto;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Purchase Requisition lifecycle (ADR-0027 D-3, FR-PROC-01..05).
 */
public interface PurchaseRequisitionService {

    PurchaseRequisitionDto create(CreatePurchaseRequisitionRequest req);

    PurchaseRequisitionDto getByUid(String uid);

    Page<PurchaseRequisitionDto> list(Long companyId, Pageable pageable);

    /** DRAFT → SUBMITTED; assigns requisition_number (PR-####). */
    PurchaseRequisitionDto submit(String uid);

    /** SUBMITTED → APPROVED (permission-gated fallback). */
    PurchaseRequisitionDto approve(String uid);

    /** SUBMITTED → REJECTED (permission-gated fallback). */
    PurchaseRequisitionDto reject(String uid, String reason);

    /**
     * APPROVED → CONVERTED.
     * Creates the target document (RFQ or Purchase Order) from the requisition's lines in the same
     * transaction, sets {@code convertedToUid}/{@code convertedToType} on the requisition, and
     * returns the uid of the created document (D-3).
     */
    String convert(String uid, ConvertRequisitionRequest req);

    /** DRAFT/SUBMITTED → CANCELLED. */
    PurchaseRequisitionDto cancel(String uid, String reason);
}

File: web/src/app/features/admin/purchases/requisition/purchase-requisition.service.ts
import { HttpClient, HttpContext, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable, map } from 'rxjs';
import { ApiResponse, PageMeta } from '../../../../core/api/api-response.model';
import { SKIP_UNWRAP } from '../../../../core/api/http-context.tokens';
import { environment } from '../../../../../environments/environment';
import {
  ConvertRequisitionRequest,
  CreatePurchaseRequisitionRequest,
  PurchaseRequisitionDto,
} from './purchase-requisition.model';

export interface PurchaseRequisitionPage {
  rows: PurchaseRequisitionDto[];
  meta: PageMeta;
}

/**
 * Purchase Requisition API client.
 * list() uses SKIP_UNWRAP to read both data and PageMeta.
 * All other methods use the auto-unwrap path (interceptor strips the envelope).
 * Base: /api/v1/purchase-requisitions
 */
@Injectable({ providedIn: 'root' })
export class PurchaseRequisitionService {
  private readonly http = inject(HttpClient);
  private readonly base = `${environment.apiBaseUrl}/purchase-requisitions`;

  // ── List / Get ─────────────────────────────────────────────────────────────

  list(
    companyId: string,
    status?: string,
    page = 0,
    size = 20,
  ): Observable<PurchaseRequisitionPage> {
    let params = new HttpParams()
      .set('companyId', companyId)
      .set('page', String(page))
      .set('size', String(size));
    if (status?.trim()) params = params.set('status', status.trim());

    const context = new HttpContext().set(SKIP_UNWRAP, true);
    return this.http
      .get<ApiResponse<PurchaseRequisitionDto[]>>(this.base, { params, context })
      .pipe(
        map((env) => ({
          rows: env.data ?? [],
          meta: env.meta ?? {
            page,
            size,
            totalElements: env.data?.length ?? 0,
            totalPages: 1,
            hasNext: false,
          },
        })),
      );
  }

  getByUid(uid: string): Observable<PurchaseRequisitionDto> {
    return this.http.get<PurchaseRequisitionDto>(`${this.base}/uid/${uid}`);
  }

  // ── Create ─────────────────────────────────────────────────────────────────

  create(request: CreatePurchaseRequisitionRequest): Observable<PurchaseRequisitionDto> {
    return this.http.post<PurchaseRequisitionDto>(this.base, request);
  }

  // ── Lifecycle ──────────────────────────────────────────────────────────────

  submit(uid: string): Observable<PurchaseRequisitionDto> {
    return this.http.post<PurchaseRequisitionDto>(`${this.base}/uid/${uid}/submit`, {});
  }

  approve(uid: string): Observable<PurchaseRequisitionDto> {
    return this.http.post<PurchaseRequisitionDto>(`${this.base}/uid/${uid}/approve`, {});
  }

  /** reason passed as query param per controller: ?reason=... */
  reject(uid: string, reason: string): Observable<PurchaseRequisitionDto> {
    const params = new HttpParams().set('reason', reason);
    return this.http.post<PurchaseRequisitionDto>(
      `${this.base}/uid/${uid}/reject`,
      {},
      { params },
    );
  }

  /**
   * Converts the requisition to a PURCHASE_ORDER or RFQ, actually creating the target document
   * server-side (D-3). Body-based contract (no more ?targetType= query param): RFQ requires
   * supplierUids (non-empty); PURCHASE_ORDER requires supplierUid. Returns the created
   * document's uid (String).
   */
  convert(uid: string, request: ConvertRequisitionRequest): Observable<string> {
    return this.http.post<string>(`${this.base}/uid/${uid}/convert`, request);
  }

  cancel(uid: string, reason: string): Observable<PurchaseRequisitionDto> {
    const params = new HttpParams().set('reason', reason);
    return this.http.post<PurchaseRequisitionDto>(
      `${this.base}/uid/${uid}/cancel`,
      {},
      { params },
    );
  }
}
